File: Question 0077 - Combinations/src/Solution.java
import java.util.ArrayList;
import java.util.List;

class Solution
{
    private void solve(List<List<Integer>> ans,List<Integer> list, int n, int k, int start)
    {
        if(k==0)
        {
            ans.add(new ArrayList<>(list));
            return;
        }

        for (int i = start; i <= n; i++) {
            list.add(i);
            solve(ans,list,n,k-1,i+1);
            list.remove(list.size()-1);
        }
    }
    public List<List<Integer>> combine(int n, int k)
    {
        List<List<Integer>> ans = new ArrayList<>();
        List<Integer> list = new ArrayList<>();
        solve(ans,list,n,k, 1);
        return ans;
    }
}

File: Question 0077 - Combinations/src/Solution.cpp
#include<bits/stdc++.h>
using namespace std;

void solve(int n, int k, vector<int>& curr, vector<vector<int>>& ans, int start)
{
    if(curr.size() == k)
    {
        ans.push_back(curr);
        return;
    }

    for(int i = start; i <= n; i++)
    {
        curr.push_back(i);
        solve(n, k, curr, ans, i+1);
        curr.pop_back();
    }
}

class Solution 
{
public:
    vector<vector<int>> combine(int n, int k) 
    {
        vector<vector<int>> ans;
        vector<int> curr;
        solve(n, k, curr, ans, 1);
        return ans;
    }
};